infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/SessionManager.java

```diff
@@ -0,0 +1,362 @@
+import org.corfudb.runtime.LogReplication.LogReplicationSession;
+import org.corfudb.runtime.LogReplication.ReplicationSubscriber;
+import org.corfudb.runtime.collections.CorfuStore;
+import org.corfudb.runtime.collections.TxnContext;
+import org.corfudb.runtime.exceptions.TransactionAbortedException;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError;
+import org.corfudb.runtime.proto.RpcCommon.UuidMsg;
+import org.corfudb.util.NodeLocator;
+import org.corfudb.util.retry.IRetry;
+import org.corfudb.util.retry.IntervalRetry;
+import org.corfudb.util.retry.RetryNeededException;
+
+import javax.annotation.Nonnull;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+/**
+ * Manage log replication sessions for multiple replication models.
+ *
+ * A replication session is determined by the cluster endpoints (source & sink)--given by the
+ * Cluster Manager (topology)--the replication model and the client (given by clients upon
+ * explicit registration or implicitly through protoBuf schema options).
+ */
+@Slf4j
+public class SessionManager {
+
+    // Represents default client for LR V1, i.e., use case where tables are tagged with
+    // 'is_federated' flag, yet no client is specified in proto
+    private static final String DEFAULT_CLIENT = "00000000-0000-0000-0000-000000000000";
+
+    private static final UUID DEFAULT_CLIENT_ID = UUID.fromString(DEFAULT_CLIENT);
+
+    private final CorfuStore corfuStore;
+
+    private final CorfuRuntime runtime;
+
+    private String localCorfuEndpoint;
+
+    private CorfuReplicationManager replicationManager;
+
+    private LogReplicationConfigManager configManager;
+
+    private Set<LogReplicationSession> sessions = new HashSet<>();
+
+    private Set<LogReplicationSession> incomingSessions = new HashSet<>();
+
+    private Set<LogReplicationSession> outgoingSessions = new HashSet<>();
+
+    Map<LogReplicationSession, LogReplicationContext> sessionToContextMap = new HashMap<>();
+
+    @Getter
+    private TopologyDescriptor topology;
+
+    @Getter
+    private final LogReplicationMetadataManager metadataManager;
+
+    private final LogReplicationUpgradeManager upgradeManager;
+
+    /**
+     * Constructor
+     *
+     * @param topology            the current topology
+     * @param corfuRuntime        runtime for database access
+     * @param serverContext       the server context
+     */
+    public SessionManager(@Nonnull TopologyDescriptor topology, CorfuRuntime corfuRuntime,
+                          ServerContext serverContext, LogReplicationUpgradeManager upgradeManager) {
+        this.topology = topology;
+        this.runtime = corfuRuntime;
+        this.corfuStore = new CorfuStore(corfuRuntime);
+        this.localCorfuEndpoint = NodeLocator.parseString(serverContext.getLocalEndpoint()).getHost() + ":" +
```

Comment:
can we do this instead to be IPV6 compatible
`serverContext.getNodeLocator().toEndpointUrl();`

Reply:
there is no way to get the node locator from the server context...

Reply:
Looks like, the getter of NodeLocator is not available.

@pankti-m , can we add this like following snippet? serverConfig should be available in server context. let me know if you would like me to try this.

```
        nodeLocator = NodeLocator
                .parseString(serverConfig.get("--address") + ":" + serverConfig.get("<port>"));
        localEndpoint = nodeLocator.toEndpointUrl();
```

Reply:
Changed as discussed offline